public/app/features/explore/AddToDashboard/AddToDashboardModal.test.tsx

```diff
@@ -0,0 +1,137 @@
+import * as dashboardApi from 'app/features/manage-dashboards/state/actions';
+
+const createFolder = (title: string, id: number): DashboardSearchHit => ({
+  title,
+  id,
+  isStarred: false,
+  type: DashboardSearchItemType.DashFolder,
+  items: [],
+  url: '',
+  uri: '',
+  tags: [],
+});
+
+describe('Add to Dashboard Modal', () => {
+  const searchFoldersResponse = Promise.resolve([createFolder('Folder 1', 1), createFolder('Folder 2', 2)]);
+
+  const waitForSearchFolderResponse = async () => {
+    return act(async () => {
+      // FolderPicker asyncrounously sets its internal state based on search results, causing warnings when testing.
+      // Given we are not aware of the component implementation to wait on certain element to appear or disappear (for example a loading indicator),
+      // we wait for the mocked promise we know it internally uses.
+      // This is less than ideal as we are relying on implementation details, but is a reasonable solution for this test's scope
+      await searchFoldersResponse;
+    });
+  };
+
+  beforeEach(() => {
+    jest.spyOn(dashboardApi, 'searchFolders').mockReturnValue(searchFoldersResponse);
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  describe('Save to new dashboard', () => {
+    it('Does not submit if the form is invalid', async () => {
+      const saveMock = jest.fn();
+
+      render(<AddToDashboardModal queries={[]} visualization="table" onSave={saveMock} onClose={() => {}} />);
+
+      // there shouldn't be any alert in the modal
+      expect(screen.queryByRole('alert')).not.toBeInTheDocument();
+
+      const dashboardNameInput = screen.getByRole<HTMLInputElement>('textbox', { name: /dashboard name/i });
+
+      // dashboard name is required
+      userEvent.clear(dashboardNameInput);
+
+      userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+      // The error message should appear
+      await screen.findByRole('alert');
+
+      // Create dashboard API is not invoked
+      expect(saveMock).not.toHaveBeenCalled();
+    });
+
+    it('Correctly submits if the form is valid', async () => {
+      const saveMock = jest.fn();
+
+      render(<AddToDashboardModal queries={[]} visualization="table" onSave={saveMock} onClose={() => {}} />);
+      await waitForSearchFolderResponse();
+
+      const dashboardNameInput = screen.getByRole<HTMLInputElement>('textbox', { name: /dashboard name/i });
+
+      userEvent.click(screen.getByRole('button', { name: /save and keep exploring/i }));
+
+      await waitFor(() => {
+        expect(screen.getByRole('button', { name: /save and keep exploring/i })).toBeEnabled();
+      });
+
+      expect(saveMock).toHaveBeenCalledWith(
+        {
+          dashboardName: dashboardNameInput.value,
+          queries: [],
+          visualization: 'table',
+          folderId: 1,
+        },
+        expect.anything()
```

Comment:
Would it make sense to assert the redirect as well?

```suggestion
        false
```

Reply:
here i'm only testing if the form submits correctly, given  the navigation logic it's not handled in the tested component but in the wrapper one (`index.tesx`) is tested in https://github.com/grafana/grafana/blob/909130cde9bef0e9f937b306cbe12675fe1bdcdd/public/app/features/explore/AddToDashboard/index.test.tsx#L75